core/src/Kitty.sol

```diff
@@ -5,183 +5,259 @@ import {ERC20, SafeTransferLib} from "solmate/utils/SafeTransferLib.sol";
+        _unsafeMint(to, shares);
 
-        packedSlot = _packedSlot;
+        _save(slot0, slot1); // TODO if inventory == slot0.lastBalance, then no interest was accrued, and we don't need to write slot1 to storage
     }
 
-    function withdraw(uint256 shares) external returns (uint256 amount) {
-        // Poke (includes reentrancy guard)
-        (PackedSlot memory _packedSlot, uint256 _totalSupply, uint256 _inventory) = accrueInterest();
+    function withdraw(uint112 shares, address to) external returns (uint112 amount) {
+        (Slot0 memory slot0, Slot1 memory slot1) = _load();
+
+        // Reentrancy guard
+        require(slot0.lastAccrualTime != 0);
+        lastAccrualTime = 0;
+
+        uint256 inventory;
+        (slot0, slot1, inventory) = _accrueInterest(slot0, slot1);
 
-        amount = FullMath.mulDiv(_inventory, shares, _totalSupply);
+        amount = uint112(inventory.mulDiv(shares, slot0.totalSupply)); // TODO safe casting
         require(amount != 0, "Aloe: amount too low"); // TODO use real Error
 
         // Transfer tokens
-        asset.safeTransfer(msg.sender, amount);
+        slot0.lastBalance -= amount;
+        ASSET.safeTransfer(to, amount);
 
         // Burn shares (emits event that can be interpreted as a withdrawal)
-        _burn(msg.sender, shares);
+        _unsafeBurn(msg.sender, shares);
+        unchecked {
+            slot0.totalSupply -= shares;
+        }
 
-        packedSlot = _packedSlot;
+        _save(slot0, slot1); // TODO if inventory == slot0.lastBalance, then no interest was accrued, and we don't need to write slot1 to storage
     }
 
-    function borrow(uint128 amount) external onlyMarginAccount {
-        // Poke (includes reentrancy guard)
-        (PackedSlot memory _packedSlot, , ) = accrueInterest();
+    // TODO prevent new borrows after 2100
+    function borrow(uint256 amount, address to) external onlyMarginAccount {
+        (Slot0 memory slot0, Slot1 memory slot1) = _load();
+
+        // Reentrancy guard
+        require(slot0.lastAccrualTime != 0);
+        lastAccrualTime = 0;
+
+        (slot0, slot1, ) = _accrueInterest(slot0, slot1);
 
-        borrows[msg.sender] += FullMath.mulDiv(1e18, amount, _packedSlot.borrowIndex);
-        _packedSlot.totalBorrows += amount;
+        uint256 base = amount.mulDivRoundingUp(BORROWS_SCALER, slot1.borrowIndex);
+        borrows[msg.sender] += base;
+        slot1.borrowBase += uint184(base); // TODO safe casting
 
-        asset.safeTransfer(msg.sender, amount);
-        packedSlot = _packedSlot;
+        // Transfer tokens
+        slot0.lastBalance -= uint112(amount); // TODO safe casting
+        ASSET.safeTransfer(to, amount);
+        
+        _save(slot0, slot1);
     }
 
-    function repay(uint128 amount) external onlyMarginAccount {
-        // Poke (includes reentrancy guard)
-        (PackedSlot memory _packedSlot, , ) = accrueInterest();
+    function repay(uint256 amount, address to) external {
```

Comment:
Would it be smart to check to see if the to address is what we are expecting?

Reply:
Well for borrowing, we definitely need to make sure that only official MarginAccount instances have access. But for repaying, I think we can open it up to anyone. If you want to repay Bob's debt for him, go ahead! The protocol doesn't care. The more practical use-case is if someone's MarginAccount is about to get liquidated. Rather than going through that to pay off debt, they can call this function from **any address** that happens to have the requisite funds. This may be more gas efficient for them.